src/stan/analyze/mcmc/compute_effective_sample_size.hpp

```diff
@@ -1,98 +1,184 @@
+    size_t s = 1;
+    for (; s < (num_draws - 4) && (rho_hat_even + rho_hat_odd) > 0;
+         s += 2) {
       for (int chain = 0; chain < num_chains; ++chain)
-        acov_s(chain) = acov(chain)(1);
-      double rho_hat_even = 1;
-      double rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
-      rho_hat_s(1) = rho_hat_odd;
-      // Geyer's initial positive sequence
-      int max_s = 1;
-      for (size_t s = 1;
-           (s < (num_draws - 2) && (rho_hat_even + rho_hat_odd) >= 0);
-           s += 2) {
-        for (int chain = 0; chain < num_chains; ++chain)
-          acov_s(chain) = acov(chain)(s + 1);
-        rho_hat_even = 1 - (mean_var - acov_s.mean()) / var_plus;
-        for (int chain = 0; chain < num_chains; ++chain)
-          acov_s(chain) = acov(chain)(s + 2);
-        rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
-        if ((rho_hat_even + rho_hat_odd) >= 0) {
-          rho_hat_s(s + 1) = rho_hat_even;
-          rho_hat_s(s + 2) = rho_hat_odd;
-        }
-        max_s = s + 2;
-      }
-      // Geyer's initial monotone sequence
-      for (int s = 3; s <= max_s - 2; s += 2) {
-        if (rho_hat_s(s + 1) + rho_hat_s(s + 2) >
-            rho_hat_s(s - 1) + rho_hat_s(s)) {
-          rho_hat_s(s + 1) = (rho_hat_s(s - 1) + rho_hat_s(s)) / 2;
-          rho_hat_s(s + 2) = rho_hat_s(s + 1);
-        }
+        acov_s(chain) = acov(chain)(s + 1);
+      rho_hat_even = 1 - (mean_var - acov_s.mean()) / var_plus;
+      for (int chain = 0; chain < num_chains; ++chain)
+        acov_s(chain) = acov(chain)(s + 2);
+      rho_hat_odd = 1 - (mean_var - acov_s.mean()) / var_plus;
+      if ((rho_hat_even + rho_hat_odd) >= 0) {
+        rho_hat_s(s + 1) = rho_hat_even;
+        rho_hat_s(s + 2) = rho_hat_odd;
       }
+    }
 
-      return num_chains * num_draws / (1 + 2 * rho_hat_s.sum());
+    int max_s = s;
+    // this is used in the improved estimate
+    if (rho_hat_even > 0)
+      rho_hat_s(max_s + 1) = rho_hat_even;
+
+    // Geyer's initial monotone sequence
+    for (int s = 1; s <= max_s - 3; s += 2) {
+      if (rho_hat_s(s + 1) + rho_hat_s(s + 2) >
+          rho_hat_s(s - 1) + rho_hat_s(s)) {
+        rho_hat_s(s + 1) = (rho_hat_s(s - 1) + rho_hat_s(s)) / 2;
+        rho_hat_s(s + 2) = rho_hat_s(s + 1);
+      }
     }
 
-    inline
-    double compute_effective_sample_size(std::vector<const double*> draws,
-                                         size_t size) {
-      int num_chains = draws.size();
-      std::vector<size_t> sizes(num_chains, size);
-      return compute_effective_sample_size(draws, sizes);
+    double ess = num_chains * num_draws;
+    // Geyer's truncated estimate
+    // Improved estimate reduces variance in antithetic case
+    double tau_hat = -1 + 2*rho_hat_s.head(max_s).sum() + rho_hat_s(max_s + 1);
+    // Safety check for negative values and
+    // with max ess equal to ess*log10(ess)
+    return ess / std::max(tau_hat, 1 / std::log10(ess));
```

Comment:
I think the logic here would be much easier to read as

```
return std::max(num_total_draws / tau_hat, num_total_raws * std::log10(num_total_draws));
```

Reply:
The proposed line is not mathematically equivalent, and I couldn't come up with a better way to write it.  After considering a change here, I decided to keep it the way it is.

Reply:
Sorry, that should have been std::min rather than std::max, which in my opinion much more clearly communicates that the ESS estimator is thresholded at `num_total_draws * log(num_total_draws)`.

I strongly believe that the code should be

```
double num_total_draws = num_chains * num_draws;
double tau_hat = -1 + 2*rho_hat_s.head(max_s).sum() + rho_hat_s(max_s + 1);
return std::min(num_total_draws / tau_hat, num_total_draws * std::log10(num_total_draws));
```
or even
```
double num_total_draws = num_chains * num_draws;
double tau_hat = -1 + 2*rho_hat_s.head(max_s).sum() + rho_hat_s(max_s + 1);
double ess_nominal = num_total_draws / tau_hat;
double ess_threshold = num_total_draws * std::log10(num_total_draws);
return std::min(ess_nominal, ess_threshold);
```

Reply:
I like it.  Done.